### PR TITLE
Handle extra info in dylink section

### PR DESCRIPTION
**Pull request by sbc100**
If extra data is found in this section simply propagate it.

Also, remove some dead code from wasm-binary.cpp.

### CODE REVIEW

File: src/wasm.h [hunk 1]
```diff
@@ -1824,6 +1824,7 @@ class UserSection {
 class DylinkSection {
 public:
   Index memorySize, memoryAlignment, tableSize, tableAlignment;
+  std::vector<char> tail;
```
Comment (kripken) [inline src/wasm.h:1827]: As this is written after the neededDynlibs, how about moving it a line down?

**Review by kripken — APPROVED** [review]
What kind of data appears here? (I've probably missed some recent LLVM work for this.)

For a test, maybe roundtripping a binary?

File: src/wasm.h (hunk 1 above)
Comment (sbc100) [reply src/wasm.h:1827]: > What kind of data appears here? (I've probably missed some recent LLVM work for this.)

Nothing yes, but I'm adding more the moment, and I would like to just work without having to teach binaryen about it (at least not right away).  This allows me to experiment with adding more.

> 
> For a test, maybe roundtripping a binary?

Yup, will add.

Comment (sbc100) [reply src/wasm.h:1827]: Where would be good place in binaryen add a binary round-trip test like that?

Comment (kripken) [reply src/wasm.h]: Right now the best we have is `tests/passes/` where you can put a `.wasm` and a `.passes` combination. The passes can include `--roundtrip --print`.

Comment (tlively) [reply src/wasm.h]: I would prefer adding a `test/lit/binary` directory with a `mytest.test` file and `mytest.wasm` file, where `mytest.test` contains RUN lines for performing the test.

File: src/wasm-binary.h [hunk 2]
```diff
@@ -1251,11 +1252,7 @@ class WasmBinaryWriter {
       : data(data), size(size), pointerLocation(pointerLocation) {}
   };
 
-  std::vector<Buffer> buffersToWrite;
-
   void emitBuffer(const char* data, size_t size);
```
Comment (tlively) [inline src/wasm-binary.h:1255]: Can we delete `emitBuffer` and `buffersToWrite`, too? If not, why can we delete `finishUp`?

Comment (sbc100) [reply src/wasm-binary.h:1255]: buffersToWrite is deleted right?   (I forgot to remove the declaration of `emitBuffer`.. will do that now).

File: src/wasm.h (hunk 1 above)
Comment (sbc100) [reply src/wasm.h]: Added test.  Thanks!

File: scripts/binaryen-lit.in [hunk 3]
```diff
@@ -27,7 +27,7 @@ builtin_parameters = {
     }
 }
 
-print(builtin_parameters)
+#print(builtin_parameters)
```
Comment (kripken) [inline scripts/binaryen-lit.in:30]: intentional?

File: test/lit/binary/dylink.test [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Verify that extra bytes in the dylink section are preserved.
+#
+# `dylink.test.wasm` was generated using emscripten to build side
```
Comment (kripken) [inline test/lit/binary/dylink.test:3]:
```suggestion
# `dylink.test.wasm` was generated using emscripten to build a side
```

File: test/lit/binary/dylink.test [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+# module with two TLS exports (which adds extra information to the
+# dylink section) using:
+#  `emcc -s SIDE_MODULE side.c`
+# and then removing all sections except the dylink section using:
+#  `llvm-objcopy --only-section=dylink`
+RUN: wasm-opt -O1 %s.wasm -o %t.o
+RUN: cmp %s.wasm %t.o
```
Comment (kripken) [inline test/lit/binary/dylink.test:10]: Is the source file short enough to quote here in the comment?

File: scripts/binaryen-lit.in (hunk 3 above)
Comment (sbc100) [reply scripts/binaryen-lit.in:30]: This was extra debugging information .. I should probably just completely remove it.

File: test/lit/binary/dylink.test (hunk 5 above)
Comment (sbc100) [reply test/lit/binary/dylink.test:10]: At yes I can do that.

**Review by kripken — APPROVED** [review]

**Pull request merged**